File: app/src/main/java/com/anokmik/tripassistant/databinding/list/RecyclerViewListChangedCallback.java
package com.anokmik.tripassistant.databinding.list;

import android.databinding.ObservableList;
import android.support.v7.widget.RecyclerView;

public final class RecyclerViewListChangedCallback extends ObservableList.OnListChangedCallback {

    private final RecyclerView.Adapter adapter;

    public RecyclerViewListChangedCallback(RecyclerView.Adapter adapter) {
        this.adapter = adapter;
    }

    @Override
    public void onChanged(ObservableList sender) {
        adapter.notifyDataSetChanged();
    }

    @Override
    public void onItemRangeChanged(ObservableList sender, int positionStart, int itemCount) {
        adapter.notifyItemRangeChanged(positionStart, itemCount);
    }

    @Override
    public void onItemRangeInserted(ObservableList sender, int positionStart, int itemCount) {
        adapter.notifyItemRangeInserted(positionStart, itemCount);
    }

    @Override
    public void onItemRangeMoved(ObservableList sender, int fromPosition, int toPosition, int itemCount) {
        for (int i = 0; i < itemCount; i++) {
            adapter.notifyItemMoved(fromPosition + i, toPosition + i);
        }
    }

    @Override
    public void onItemRangeRemoved(ObservableList sender, int positionStart, int itemCount) {
        adapter.notifyItemRangeRemoved(positionStart, itemCount);
    }

}

File: app/src/main/kotlin/com/anokmik/tripassistant/databinding/list/RecyclerViewListChangedCallback.kt
package com.anokmik.tripassistant.databinding.list

import android.databinding.ObservableList
import android.support.v7.widget.RecyclerView

class RecyclerViewListChangedCallback<T>(private val adapter: RecyclerView.Adapter<out RecyclerView.ViewHolder>) : ObservableList.OnListChangedCallback<ObservableList<T>>() {

    override fun onChanged(sender: ObservableList<T>?) = adapter.notifyDataSetChanged()

    override fun onItemRangeInserted(sender: ObservableList<T>?, positionStart: Int, itemCount: Int) = adapter.notifyItemRangeChanged(positionStart, itemCount)

    override fun onItemRangeChanged(sender: ObservableList<T>?, positionStart: Int, itemCount: Int) = adapter.notifyItemRangeInserted(positionStart, itemCount)

    override fun onItemRangeMoved(sender: ObservableList<T>?, fromPosition: Int, toPosition: Int, itemCount: Int) {
        for (i in 0..itemCount - 1) {
            adapter.notifyItemMoved(fromPosition + i, toPosition + i)
        }
    }

    override fun onItemRangeRemoved(sender: ObservableList<T>?, positionStart: Int, itemCount: Int) = adapter.notifyItemRangeRemoved(positionStart, itemCount)

}